File: Platform/infra/security/factory/js/auth-client-factory.js
/**
 * Authentication Client Factory for JavaScript
 * 
 * This module provides factory methods and builder pattern for creating authentication clients
 * in the CoyoteSense platform. Supports multiple authentication standards:
 * - OAuth2 Client Credentials (RFC 6749)
 * - OAuth2 Authorization Code (RFC 6749) 
 * - OAuth2 + PKCE (RFC 7636)
 * - JWT Bearer (RFC 7523)
 * - mTLS Client Credentials (RFC 8705)
 */

const { 
    AuthMode, 
    AuthConfigHelper,
    InMemoryTokenStorage,
    ConsoleAuthLogger 
} = require('../../interfaces/js/auth-interfaces');

/**
 * Builder pattern for creating authentication clients with fluent interface
 */
class AuthClientBuilder {
    constructor() {
        this._config = {};
        this._tokenStorage = null;
        this._logger = null;
    }

    /**
     * Set authentication server URL
     * @param {string} url - The authentication server URL
     * @returns {AuthClientBuilder}
     */
    serverUrl(url) {
        this._config.serverUrl = url;
        return this;
    }

    /**
     * Set client credentials (ID and secret)
     * @param {string} clientId - The client ID
     * @param {string} [clientSecret] - The client secret (optional for public clients)
     * @returns {AuthClientBuilder}
     */
    clientCredentials(clientId, clientSecret) {
        this._config.clientId = clientId;
        if (clientSecret) {
            this._config.clientSecret = clientSecret;
        }
        return this;
    }

    /**
     * Set authentication mode
     * @param {string} mode - The authentication mode
     * @returns {AuthClientBuilder}
     */
    authMode(mode) {
        this._config.authMode = mode;
        return this;
    }

    /**
     * Set default scopes
     * @param {string[]} scopes - Default scopes to request
     * @returns {AuthClientBuilder}
     */
    defaultScopes(scopes) {
        this._config.defaultScopes = scopes;
        return this;
    }

    /**
     * Set mTLS certificates for mTLS authentication
     * @param {string} certPath - Path to client certificate
     * @param {string} keyPath - Path to client private key
     * @param {string} [caCertPath] - Path to CA certificate (optional)
     * @returns {AuthClientBuilder}
     */
    mtlsCertificates(certPath, keyPath, caCertPath) {
        this._config.clientCertPath = certPath;
        this._config.clientKeyPath = keyPath;
        if (caCertPath) {
            this._config.caCertPath = caCertPath;
        }
        return this;
    }

    /**
     * Set JWT settings for JWT Bearer authentication
     * @param {string} signingKeyPath - Path to JWT signing key
     * @param {string} issuer - JWT issuer
     * @param {string} audience - JWT audience
     * @param {string} [algorithm='RS256'] - JWT algorithm
     * @returns {AuthClientBuilder}
     */
    jwtSettings(signingKeyPath, issuer, audience, algorithm = 'RS256') {
        this._config.jwtSigningKeyPath = signingKeyPath;
        this._config.jwtIssuer = issuer;
        this._config.jwtAudience = audience;
        this._config.jwtAlgorithm = algorithm;
        return this;
    }

    /**
     * Set redirect URI for Authorization Code flows
     * @param {string} uri - Redirect URI
     * @returns {AuthClientBuilder}
     */
    redirectUri(uri) {
        this._config.redirectUri = uri;
        return this;
    }

    /**
     * Enable PKCE for Authorization Code flow
     * @param {boolean} [enable=true] - Whether to enable PKCE
     * @returns {AuthClientBuilder}
     */
    enablePkce(enable = true) {
        this._config.usePkce = enable;
        return this;
    }

    /**
     * Set request timeout
     * @param {number} timeoutMs - Timeout in milliseconds
     * @returns {AuthClientBuilder}
     */
    timeout(timeoutMs) {
        this._config.timeoutMs = timeoutMs;
        return this;
    }

    /**
     * Set custom headers
     * @param {Object} headers - Custom headers
     * @returns {AuthClientBuilder}
     */
    customHeaders(headers) {
        this._config.customHeaders = { ...this._config.customHeaders, ...headers };
        return this;
    }

    /**
     * Configure automatic token refresh
     * @param {boolean} [enabled=true] - Enable auto refresh
     * @param {number} [marginSeconds=300] - Refresh margin in seconds
     * @param {number} [checkIntervalMs=60000] - Check interval in milliseconds
     * @returns {AuthClientBuilder}
     */
    autoRefresh(enabled = true, marginSeconds = 300, checkIntervalMs = 60000) {
        this._config.enableAutoRefresh = enabled;
        this._config.refreshMarginSeconds = marginSeconds;
        this._config.refreshCheckIntervalMs = checkIntervalMs;
        return this;
    }

    /**
     * Set token storage implementation
     * @param {Object} storage - Token storage implementation
     * @returns {AuthClientBuilder}
     */
    tokenStorage(storage) {
        this._tokenStorage = storage;
        return this;
    }

    /**
     * Set logger implementation
     * @param {Object} logger - Logger implementation
     * @returns {AuthClientBuilder}
     */
    logger(logger) {
        this._logger = logger;
        return this;
    }

    /**
     * Build the authentication client
     * @param {Object} [httpClient] - HTTP client instance
     * @returns {Object} Authentication client
     */
    build(httpClient) {
        if (!this._config.serverUrl) {
            throw new Error('Server URL is required');
        }
        if (!this._config.clientId) {
            throw new Error('Client ID is required');
        }

        if (!httpClient) {
            throw new Error('HTTP client must be provided in build() method');
        }

        const finalConfig = {
            authMode: this._config.authMode || AuthMode.CLIENT_CREDENTIALS,
            serverUrl: this._config.serverUrl,
            clientId: this._config.clientId,
            clientSecret: this._config.clientSecret,
            defaultScopes: this._config.defaultScopes || [],
            timeoutMs: this._config.timeoutMs || 30000,
            customHeaders: this._config.customHeaders || {},
            enableAutoRefresh: this._config.enableAutoRefresh !== undefined ? this._config.enableAutoRefresh : true,
            refreshMarginSeconds: this._config.refreshMarginSeconds || 300,
            refreshCheckIntervalMs: this._config.refreshCheckIntervalMs || 60000,
            // Include all other configuration properties
            clientCertPath: this._config.clientCertPath,
            clientKeyPath: this._config.clientKeyPath,
            caCertPath: this._config.caCertPath,
            jwtSigningKeyPath: this._config.jwtSigningKeyPath,
            jwtAlgorithm: this._config.jwtAlgorithm,
            jwtIssuer: this._config.jwtIssuer,
            jwtAudience: this._config.jwtAudience,
            redirectUri: this._config.redirectUri,
            usePkce: this._config.usePkce,
            refreshBufferSeconds: this._config.refreshBufferSeconds,
            autoRefresh: this._config.autoRefresh,
            maxRetryAttempts: this._config.maxRetryAttempts,
            retryDelayMs: this._config.retryDelayMs,
            verifySsl: this._config.verifySsl
        };

        // Import the actual client implementation
        const { AuthClient } = require('../../clients/javascript/auth-client');

        return new AuthClient(
            finalConfig,
            httpClient,
            this._tokenStorage,
            this._logger
        );
    }
}

/**
 * Factory class for creating authentication clients with different authentication modes
 */
class AuthClientFactory {
    /**
     * Create a new authentication client builder
     * @returns {AuthClientBuilder}
     */
    static create() {
        return new AuthClientBuilder();
    }

    /**
     * Create authentication client with minimal configuration
     * @param {string} serverUrl - Authentication server URL
     * @param {string} clientId - Client ID
     * @param {string} [clientSecret] - Client secret
     * @param {Object} [httpClient] - HTTP client instance
     * @returns {Object} Authentication client
     */
    static createSimple(serverUrl, clientId, clientSecret, httpClient) {
        return AuthClientFactory.create()
            .serverUrl(serverUrl)
            .clientCredentials(clientId, clientSecret)
            .build(httpClient);
    }

    /**
     * Create authentication client for Client Credentials flow
     * @param {string} serverUrl - Authentication server URL
     * @param {string} clientId - Client ID
     * @param {string} clientSecret - Client secret
     * @param {string[]} [scopes] - Default scopes
     * @param {Object} [httpClient] - HTTP client instance
     * @returns {Object} Authentication client
     */
    static createClientCredentials(serverUrl, clientId, clientSecret, scopes, httpClient) {
        const builder = AuthClientFactory.create()
            .serverUrl(serverUrl)
            .clientCredentials(clientId, clientSecret)
            .authMode(AuthMode.CLIENT_CREDENTIALS);

        if (scopes) {
            builder.defaultScopes(scopes);
        }

        return builder.build(httpClient);
    }

    /**
     * Create authentication client for Client Credentials with mTLS flow
     * @param {string} serverUrl - Authentication server URL
     * @param {string} clientId - Client ID
     * @param {string} clientCertPath - Path to client certificate
     * @param {string} clientKeyPath - Path to client private key
     * @param {string} [caCertPath] - Path to CA certificate
     * @param {string[]} [scopes] - Default scopes
     * @param {Object} [httpClient] - HTTP client instance
     * @returns {Object} Authentication client
     */
    static createClientCredentialsMtls(serverUrl, clientId, clientCertPath, clientKeyPath, caCertPath, scopes, httpClient) {
        const builder = AuthClientFactory.create()
            .serverUrl(serverUrl)
            .clientCredentials(clientId)
            .authMode(AuthMode.CLIENT_CREDENTIALS_MTLS)
            .mtlsCertificates(clientCertPath, clientKeyPath, caCertPath);

        if (scopes) {
            builder.defaultScopes(scopes);
        }

        return builder.build(httpClient);
    }

    /**
     * Create authentication client for JWT Bearer flow
     * @param {string} serverUrl - Authentication server URL
     * @param {string} clientId - Client ID
     * @param {string} jwtSigningKeyPath - Path to JWT signing key
     * @param {string} jwtIssuer - JWT issuer
     * @param {string} jwtAudience - JWT audience
     * @param {string} [jwtAlgorithm='RS256'] - JWT algorithm
     * @param {string[]} [scopes] - Default scopes
     * @param {Object} [httpClient] - HTTP client instance
     * @returns {Object} Authentication client
     */
    static createJwtBearer(serverUrl, clientId, jwtSigningKeyPath, jwtIssuer, jwtAudience, jwtAlgorithm = 'RS256', scopes, httpClient) {
        const builder = AuthClientFactory.create()
            .serverUrl(serverUrl)
            .clientCredentials(clientId)
            .authMode(AuthMode.JWT_BEARER)
            .jwtSettings(jwtSigningKeyPath, jwtIssuer, jwtAudience, jwtAlgorithm);

        if (scopes) {
            builder.defaultScopes(scopes);
        }

        return builder.build(httpClient);
    }

    /**
     * Create authentication client for Authorization Code flow
     * @param {string} serverUrl - Authentication server URL
     * @param {string} clientId - Client ID
     * @param {string} redirectUri - Redirect URI
     * @param {string[]} [scopes] - Default scopes
     * @param {Object} [httpClient] - HTTP client instance
     * @returns {Object} Authentication client
     */
    static createAuthorizationCode(serverUrl, clientId, redirectUri, scopes, httpClient) {
        const builder = AuthClientFactory.create()
            .serverUrl(serverUrl)
            .clientCredentials(clientId) // No secret for public clients
            .authMode(AuthMode.AUTHORIZATION_CODE)
            .redirectUri(redirectUri);

        if (scopes) {
            builder.defaultScopes(scopes);
        }

        return builder.build(httpClient);
    }

    /**
     * Create authentication client for Authorization Code with PKCE flow
     * @param {string} serverUrl - Authentication server URL
     * @param {string} clientId - Client ID
     * @param {string} redirectUri - Redirect URI
     * @param {string[]} [scopes] - Default scopes
     * @param {Object} [httpClient] - HTTP client instance
     * @returns {Object} Authentication client
     */
    static createAuthorizationCodePkce(serverUrl, clientId, redirectUri, scopes, httpClient) {
        const builder = AuthClientFactory.create()
            .serverUrl(serverUrl)
            .clientCredentials(clientId) // No secret for public clients
            .authMode(AuthMode.AUTHORIZATION_CODE_PKCE)
            .redirectUri(redirectUri)
            .enablePkce(true);

        if (scopes) {
            builder.defaultScopes(scopes);
        }

        return builder.build(httpClient);
    }
}

// Legacy aliases for backward compatibility
const OAuth2AuthClientBuilder = AuthClientBuilder;
const OAuth2AuthClientFactory = AuthClientFactory;

module.exports = {
    AuthClientBuilder,
    AuthClientFactory,
    // Legacy exports
    OAuth2AuthClientBuilder,
    OAuth2AuthClientFactory
};

File: Platform/infra/security/factory/py/auth_client_factory.py
"""
Authentication Client Factory for Python

This module provides factory methods and builder pattern for creating authentication clients
in the CoyoteSense platform. Supports multiple authentication standards:
- OAuth2 Client Credentials (RFC 6749)
- OAuth2 Authorization Code (RFC 6749) 
- OAuth2 + PKCE (RFC 7636)
- JWT Bearer (RFC 7523)
- mTLS Client Credentials (RFC 8705)
"""

from typing import Optional, List, Dict, Any
from ...interfaces.py.auth_client import (
    AuthMode,
    AuthClientConfig,
    IAuthTokenStorage,
    IAuthLogger,
    IAuthClient
)
from ....http.py.http_client import HttpClient  # Assuming this exists


class AuthClientBuilder:
    """Builder pattern for creating authentication clients with fluent interface."""
    
    def __init__(self):
        self._config_data: Dict[str, Any] = {}
        self._token_storage: Optional[IAuthTokenStorage] = None
        self._logger: Optional[IAuthLogger] = None
    
    def server_url(self, url: str) -> 'AuthClientBuilder':
        """Set authentication server URL."""
        self._config_data['server_url'] = url
        return self
    
    def client_credentials(self, client_id: str, client_secret: Optional[str] = None) -> 'AuthClientBuilder':
        """Set client credentials (ID and secret)."""
        self._config_data['client_id'] = client_id
        if client_secret:
            self._config_data['client_secret'] = client_secret
        return self
    
    def auth_mode(self, mode: AuthMode) -> 'AuthClientBuilder':
        """Set authentication mode."""
        self._config_data['auth_mode'] = mode
        return self
    
    def default_scopes(self, scopes: List[str]) -> 'AuthClientBuilder':
        """Set default scopes."""
        self._config_data['default_scopes'] = scopes
        return self
    
    def mtls_certificates(self, cert_path: str, key_path: str, ca_cert_path: Optional[str] = None) -> 'AuthClientBuilder':
        """Set mTLS certificates for mTLS authentication."""
        self._config_data['client_cert_path'] = cert_path
        self._config_data['client_key_path'] = key_path
        if ca_cert_path:
            self._config_data['ca_cert_path'] = ca_cert_path
        return self
    
    def jwt_settings(self, signing_key_path: str, issuer: str, audience: str, algorithm: str = 'RS256') -> 'AuthClientBuilder':
        """Set JWT settings for JWT Bearer authentication."""
        self._config_data['jwt_signing_key_path'] = signing_key_path
        self._config_data['jwt_issuer'] = issuer
        self._config_data['jwt_audience'] = audience
        self._config_data['jwt_algorithm'] = algorithm
        return self
    
    def redirect_uri(self, uri: str) -> 'AuthClientBuilder':
        """Set redirect URI for Authorization Code flows."""
        self._config_data['redirect_uri'] = uri
        return self
    
    def enable_pkce(self, enable: bool = True) -> 'AuthClientBuilder':
        """Enable PKCE for Authorization Code flow."""
        self._config_data['use_pkce'] = enable
        return self
    
    def timeout(self, timeout_ms: int) -> 'AuthClientBuilder':
        """Set request timeout."""
        self._config_data['timeout_ms'] = timeout_ms
        return self
    
    def custom_headers(self, headers: Dict[str, str]) -> 'AuthClientBuilder':
        """Set custom headers."""
        self._config_data['custom_headers'] = headers
        return self
    
    def auto_refresh(self, enable: bool = True, margin_seconds: int = 300) -> 'AuthClientBuilder':
        """Set automatic token refresh settings."""
        self._config_data['enable_auto_refresh'] = enable
        self._config_data['refresh_margin_seconds'] = margin_seconds
        return self
    
    def token_storage(self, storage: IAuthTokenStorage) -> 'AuthClientBuilder':
        """Set token storage implementation."""
        self._token_storage = storage
        return self
    
    def logger(self, logger: IAuthLogger) -> 'AuthClientBuilder':
        """Set logger implementation."""
        self._logger = logger
        return self
    
    def build(self, http_client: Optional[HttpClient] = None) -> IAuthClient:
        """Build the authentication client."""
        # Validate required fields
        if not self._config_data.get('server_url'):
            raise ValueError('Server URL is required')
        if not self._config_data.get('client_id'):
            raise ValueError('Client ID is required')
        
        # Create configuration
        config = AuthClientConfig(
            auth_mode=self._config_data.get('auth_mode', AuthMode.CLIENT_CREDENTIALS),
            server_url=self._config_data['server_url'],
            client_id=self._config_data['client_id'],
            client_secret=self._config_data.get('client_secret'),
            default_scopes=self._config_data.get('default_scopes', []),
            client_cert_path=self._config_data.get('client_cert_path'),
            client_key_path=self._config_data.get('client_key_path'),
            ca_cert_path=self._config_data.get('ca_cert_path'),
            jwt_signing_key_path=self._config_data.get('jwt_signing_key_path'),
            jwt_algorithm=self._config_data.get('jwt_algorithm'),
            jwt_issuer=self._config_data.get('jwt_issuer'),
            jwt_audience=self._config_data.get('jwt_audience'),
            redirect_uri=self._config_data.get('redirect_uri'),
            use_pkce=self._config_data.get('use_pkce', False),
            timeout_ms=self._config_data.get('timeout_ms', 30000),
            custom_headers=self._config_data.get('custom_headers', {}),
            auto_refresh=self._config_data.get('auto_refresh', True),
            refresh_buffer_seconds=self._config_data.get('refresh_buffer_seconds', 300),
            max_retry_attempts=self._config_data.get('max_retry_attempts', 3),
            retry_delay_ms=self._config_data.get('retry_delay_ms', 1000),
            verify_ssl=self._config_data.get('verify_ssl', True)
        )
        
        # Validate configuration
        if not config.is_valid():
            raise ValueError('Invalid authentication configuration for the selected authentication mode')
        
        # Use provided HTTP client or create a default one
        if not http_client:
            raise ValueError('HTTP client must be provided in build() method')
        
        # Import the actual client implementation
        from ...clients.python.auth_client import AuthClientImpl
        
        return AuthClientImpl(
            config=config,
            http_client=http_client,
            token_storage=self._token_storage,
            logger=self._logger
        )


class AuthClientFactory:
    """Factory class for creating authentication clients with different authentication modes."""
    
    @staticmethod
    def create() -> AuthClientBuilder:
        """Create a new authentication client builder."""
        return AuthClientBuilder()
    
    @staticmethod
    def create_simple(
        server_url: str,
        client_id: str,
        client_secret: Optional[str] = None,
        http_client: Optional[HttpClient] = None
    ) -> IAuthClient:
        """Create authentication client with minimal configuration."""
        return (AuthClientFactory.create()
                .server_url(server_url)
                .client_credentials(client_id, client_secret)
                .build(http_client))
    
    @staticmethod
    def create_client_credentials(
        server_url: str,
        client_id: str,
        client_secret: str,
        scopes: Optional[List[str]] = None,
        http_client: Optional[HttpClient] = None
    ) -> IAuthClient:
        """Create authentication client for Client Credentials flow."""
        builder = (AuthClientFactory.create()
                  .server_url(server_url)
                  .client_credentials(client_id, client_secret)
                  .auth_mode(AuthMode.CLIENT_CREDENTIALS))
        
        if scopes:
            builder.default_scopes(scopes)
        
        return builder.build(http_client)
    
    @staticmethod
    def create_client_credentials_mtls(
        server_url: str,
        client_id: str,
        client_cert_path: str,
        client_key_path: str,
        ca_cert_path: Optional[str] = None,
        scopes: Optional[List[str]] = None,
        http_client: Optional[HttpClient] = None
    ) -> IAuthClient:
        """Create authentication client for Client Credentials with mTLS flow."""
        builder = (AuthClientFactory.create()
                  .server_url(server_url)
                  .client_credentials(client_id)
                  .auth_mode(AuthMode.CLIENT_CREDENTIALS_MTLS)
                  .mtls_certificates(client_cert_path, client_key_path, ca_cert_path))
        
        if scopes:
            builder.default_scopes(scopes)
        
        return builder.build(http_client)
    
    @staticmethod
    def create_jwt_bearer(
        server_url: str,
        client_id: str,
        jwt_signing_key_path: str,
        jwt_issuer: str,
        jwt_audience: str,
        jwt_algorithm: str = 'RS256',
        scopes: Optional[List[str]] = None,
        http_client: Optional[HttpClient] = None
    ) -> IAuthClient:
        """Create authentication client for JWT Bearer flow."""
        builder = (AuthClientFactory.create()
                  .server_url(server_url)
                  .client_credentials(client_id)
                  .auth_mode(AuthMode.JWT_BEARER)
                  .jwt_settings(jwt_signing_key_path, jwt_issuer, jwt_audience, jwt_algorithm))
        
        if scopes:
            builder.default_scopes(scopes)
        
        return builder.build(http_client)
    
    @staticmethod
    def create_authorization_code(
        server_url: str,
        client_id: str,
        redirect_uri: str,
        scopes: Optional[List[str]] = None,
        http_client: Optional[HttpClient] = None
    ) -> IAuthClient:
        """Create authentication client for Authorization Code flow."""
        builder = (AuthClientFactory.create()
                  .server_url(server_url)
                  .client_credentials(client_id)  # No secret for public clients
                  .auth_mode(AuthMode.AUTHORIZATION_CODE)
                  .redirect_uri(redirect_uri))
        
        if scopes:
            builder.default_scopes(scopes)
        
        return builder.build(http_client)
    
    @staticmethod
    def create_authorization_code_pkce(
        server_url: str,
        client_id: str,
        redirect_uri: str,
        scopes: Optional[List[str]] = None,
        http_client: Optional[HttpClient] = None
    ) -> IAuthClient:
        """Create authentication client for Authorization Code with PKCE flow."""
        builder = (AuthClientFactory.create()
                  .server_url(server_url)
                  .client_credentials(client_id)  # No secret for public clients
                  .auth_mode(AuthMode.AUTHORIZATION_CODE_PKCE)
                  .redirect_uri(redirect_uri)
                  .enable_pkce(True))
        
        if scopes:
            builder.default_scopes(scopes)
        
        return builder.build(http_client)


# Legacy aliases for backward compatibility
OAuth2AuthClientBuilder = AuthClientBuilder
OAuth2AuthClientFactory = AuthClientFactory
